src/lib.rs

```diff
@@ -172,21 +214,21 @@ where
     }
 }
 
-struct BatchState<Key: Eq + Hash, Item, T> {
+struct BatchState<Item, T> {
     items: Vec<Item>,
-    senders: Vec<Sender<Key, Item, T>>,
-    lock: Arc<tokio::sync::Mutex<()>>,
+    senders: Vec<Sender<T>>,
+    lock: Arc<Semaphore>,
```

Comment:
`tokio::Mutex<()>` is a wrapper around `Semaphore` anyway, so we might as well use the API that better fits this abstraction.

Why do I think semaphore is better for this model? We are essentially saying that only 1 worker can run a batch at a time. This is a just a way of saying we have 1 worker permit available.

In theory, we could allow having higher permit numbers and therefore more batch workers for applications that want smaller batch sizes

Reply:
Currently the point of the library is the batch concurrent work into sequential batches. I think I'd prefer to keep using `tokio::Mutex` because it's a simpler abstraction (whether it wraps `Semaphore` internally or not) that makes the code easier to read, see mutex and we know exactly how it behaves, I feel we can switch to `Semaphore` if/when we actually want to have concurrent batching.

Actually I'm not sure why we would have thread-concurrent batching, its harder to understand and for an async lib you can already do your batch concurrently async-style.